core/common/src/files/FileSystem.kt

```diff
@@ -0,0 +1,147 @@
+     * @throws kotlinx.io.IOException when the creation of one of the directories fails.
+     */
+    public fun createDirectories(path: Path, mustCreate: Boolean = false)
+
+    /**
+     * Atomically renames [source] to [destination] overriding [destination] if it already exists.
+     *
+     * When the file system does not support atomic move of [source] and [destination] corresponds to different
+     * file systems and the operation could not be performed atomically,
+     * [UnsupportedOperationException] is thrown.
+     *
+     * @param source the path to rename.
+     * @param destination desired path name.
+     *
+     * @throws kotlinx.io.files.FileNotFoundException when the [source] does not exist.
+     * @throws kotlinx.io.IOException when the move failed.
+     * @throws kotlin.UnsupportedOperationException when the file system does not support atomic move.
+     */
+    public fun atomicMove(source: Path, destination: Path)
+
+    /**
+     * Returns [Source] to read from a file represented by the [path].
+     *
+     * How a source will read the data is implementation-specific and failures caused
+     * by the missing file or, for example, lack of permissions may not be reported immediately,
+     * but postponed until the source will try to fetch data.
+     *
+     * @param path the path to read from.
+     *
+     * @throws kotlinx.io.files.FileNotFoundException when the file does not exist.
+     * @throws kotlinx.io.IOException when it's not possible to open the file for reading.
+     */
+    public fun read(path: Path): Source
+
+    /**
+     * Returns [Sink] to write into a file represented by the [path].
+     * File will be created if it does not exist yet.
+     *
+     * How a sink will write the data is implementation-specific and failures caused,
+     * for example, by the lack of permissions may not be reported immediately,
+     * but postponed until the sink will try to store data.
+     *
+     * @throws kotlinx.io.IOException when it's not possible to open the file for writing.
+     */
+    public fun write(path: Path): Sink
+
+    /**
+     * Return [FileMetadata] associated with a file or directory represented by the [path].
+     * If there is no such file or directory, or it's impossible to fetch metadata,
+     * `null` is returned.
+     *
+     * @param path the path to get the metadata for.
+     */
+    public fun metadataOrNull(path: Path): FileMetadata?
+
+    public companion object {
+        /**
+         * An instance of [FileSystem] representing a default system-wide file system.
+         */
+        public val System: FileSystem = SystemFileSystem
+    }
+}
+
+internal abstract class SystemFileSystemImpl : FileSystem {
+    @Suppress("DEPRECATION")
+    override fun read(path: Path): Source = path.source()
+
+    @Suppress("DEPRECATION")
+    override fun write(path: Path): Sink = path.sink()
+}
+
+internal expect val SystemFileSystem: FileSystem
+
+public class FileMetadata(
```

Comment:
This class is missing a root-level KDoc

Reply:
Trying to cut edges with "the package is more experimental than the whole library" in mind, sorry. Will add the docs.

Reply:
Fixed